### DISCUSSION TITLE
Raspberry 5 custom PCB

### DISCUSSION THREAD

**Discussion by danic85** [makerforgetech/modular-biped#64] [Ideas]
Now that I'm upgrading the project to utilise the Raspberry Pi 5, I want to also take the opportunity to upgrade the custom PCB in the head.

There is a challenge since the new pi uses an active cooling fan which gets in the way of the current PCB, but it also gives me the opportunity to incorporate some of the newer modules, including I2C neopixel support and audio output (as the Pi 5 does not have an audio jack, this may need to be I2S).

### Current approach

I currently have a PCB that connects to the 40 pin header of a Raspberry PI, and I'm looking to expand the functionality as well as upgrade to the new Raspberry Pi 5.

The PCB currently allows the following modules to be connected to the Raspberry Pi via the 40 pin header:

- Adafruit I2S MEMS Microphone Breakout - SPH0645LM4H (https://www.adafruit.com/product/3421) to allow microphone input via I2S
- Adafruit I2S Amplifier (https://shop.pimoroni.com/products/adafruit-i2s-amplifier-bff-add-on-for-qt-py-and-xiao?variant=41033041772627) to allow speaker output via I2S (new, untested)
- Adafruit NeoDriver - I2C to NeoPixel Driver Board (https://thepihut.com/products/adafruit-neodriver-i2c-to-neopixel-driver-board-stemma-qt?variant=42466863907011) to allow control of neopixels via I2C  
- Logic Level Converter (https://thepihut.com/products/sparkfun-logic-level-converter-bi-directional) to convert the 3v3 of the pi to 5v for connection to an arduino board
- RCWL-0516 Microwave motion sensor (https://hobbycomponents.com/sensors/899-rcwl-0516-microwave-radar-motion-detector-) to detect human presence
- A buzzer for sound output
- A few headers to supply power or enable a serial connection, or control a small servo (new).

Currently the custom PCB allows users to solder these modules to the PCB, then connect the PCB to the Raspberry Pi. The problem is that if the layout of any of these modules changes, or if they are out of stock, they can't be easily swapped for another module so we need to design a new PCB each time.

Therefore, the plan is to create a new PCB that will allow readily available modules to be incorporated, but then I'd also like to attempt to create a custom PCB that includes the functionality from these modules directly as integrated components, possibly using a PCBA service to assemble the boards.

I'm in discussions with a PCB service to collaborate on this, but anyone in the community has any experience and wants to help out, please let me know.

**Comment by danic85** [comment]
I'm currently using Fritzing to design the boards, but I'd be interested to hear what other people use for this (preferably free to use, as I want it to be accessible to others).

**Reply by WhoIsMrSentry** [reply]
I have used EasyEDA, KİCAD and Autodesk Eagle free before, they are a bit complicated but ideal for professional level printed circuit boards.

**Comment by WhoIsMrSentry** [comment]
> Now that I'm upgrading the project to utilise the Raspberry Pi 5, I want to also take the opportunity to upgrade the custom PCB in the head.
> 
> There is a challenge since the new pi uses an active cooling fan which gets in the way of the current PCB, but it also gives me the opportunity to incorporate some of the newer modules, including I2C neopixel support and audio output (as the Pi 5 does not have an audio jack, this may need to be I2S).
> 
> ### Current approach
> I currently have a PCB that connects to the 40 pin header of a Raspberry PI, and I'm looking to expand the functionality as well as upgrade to the new Raspberry Pi 5.
> 
> The PCB currently allows the following modules to be connected to the Raspberry Pi via the 40 pin header:
> 
> * Adafruit I2S MEMS Microphone Breakout - SPH0645LM4H (https://www.adafruit.com/product/3421) to allow microphone input via I2S
> * Adafruit I2S Amplifier (https://shop.pimoroni.com/products/adafruit-i2s-amplifier-bff-add-on-for-qt-py-and-xiao?variant=41033041772627) to allow speaker output via I2S (new, untested)
> * Adafruit NeoDriver - I2C to NeoPixel Driver Board (https://thepihut.com/products/adafruit-neodriver-i2c-to-neopixel-driver-board-stemma-qt?variant=42466863907011) to allow control of neopixels via I2C
> * Logic Level Converter (https://thepihut.com/products/sparkfun-logic-level-converter-bi-directional) to convert the 3v3 of the pi to 5v for connection to an arduino board
> * RCWL-0516 Microwave motion sensor (https://hobbycomponents.com/sensors/899-rcwl-0516-microwave-radar-motion-detector-) to detect human presence
> * A buzzer for sound output
> * A few headers to supply power or enable a serial connection, or control a small servo (new).
> 
> Currently the custom PCB allows users to solder these modules to the PCB, then connect the PCB to the Raspberry Pi. The problem is that if the layout of any of these modules changes, or if they are out of stock, they can't be easily swapped for another module so we need to design a new PCB each time.
> 
> Therefore, the plan is to create a new PCB that will allow readily available modules to be incorporated, but then I'd also like to attempt to create a custom PCB that includes the functionality from these modules directly as integrated components, possibly using a PCBA service to assemble the boards.
> 
> I'm in discussions with a PCB service to collaborate on this, but anyone in the community has any experience and wants to help out, please let me know.


If you need help with anything, I can help you. I have very little knowledge about PCB but I can improve myself

**Comment by Chompalot** [comment]
Sounds an ambitious, but logical step. Unfortunately it's been 20 years or so since I was laying out PCB's so I dont have access to the software I used to have at work  (Visula -  Zucan) the pcb's were populated by other departments. The only pcb's I did for myself were for the motor controllers for a robot.
I look forward to seeing progress :)

**Comment by danic85** [comment]
A couple of updates in the new `v4` branch under `circuits/v4` using KiCad:

![Screenshot from 2023-11-02 13-35-39](https://github.com/makerforgetech/modular-biped/assets/6583012/11444401-6de0-4b97-8fa0-82cbc80bc5d3)

![Screenshot from 2023-11-02 08-49-33](https://github.com/makerforgetech/modular-biped/assets/6583012/90a8c3e4-009b-4431-991c-4b90f12083a4)



I created custom symbols and footprints for the modules, and added them to libraries so that they can be reused. The footprints aren't correct yet, but it's a start. I'm learning the tool as I go, so expect errors ;)

The goal is to have a PCB that can either connect at 90 degrees from the pi (down the side), or via a 40 pin connector cable. But I may also make a version that connects via a hat, I just don't want to block the fan.

The new head is going to have a lot of the circuitry exposed as I think it looks interesting and allows me to talk through the construction during demos, rather than it all being hidden behind the frame. I'll see how well I can lean in to that as I finalise this PCB.

Update: added SMD neopixels so that we can show status indicators for the various modules, and JST connectors for a few optional breakout boards.

Update: swapped the pins around so that the servo and extension board slots use PWM pins (better for servo control)

**Comment by danic85** [comment]
Updated the footprints. I need to check the layers are correct then do the routing.

![Screenshot from 2023-11-10 21-36-10](https://github.com/makerforgetech/modular-biped/assets/6583012/aad1b84f-1697-4896-8942-b2f5633b7b7d)

Reference:
JST XH 2.5MM pitch
Headers 2.54MM pitch
https://cdn-learn.adafruit.com/downloads/pdf/adafruit-max98357-i2s-class-d-mono-amp.pdf
https://www.makerhero.com/img/files/download/RCWL0516-Datasheet.pdf
https://cdn-learn.adafruit.com/downloads/pdf/adafruit-i2s-mems-microphone-breakout.pdf
https://learn.adafruit.com/adafruit-neodriver-i2c-to-neopixel-driver/downloads

**Reply by danic85** [reply]
Also the pin positions need review. I can see the LLC is wrong.

**Reply by danic85** [reply]
Pin positions reviewed, pushed to branch.

**Comment by danic85** [comment]
I have the mic and SK6812s on order. Will continue testing once I have those.

I deliberately made the neopixel driver a generic header, as the one I use is now out of stock, so I am building the board so that that module can be swapped for another easily. I don't like the form factor of it anyway so this will allow flexibility in future. The others are as close to their package as I can get them.

**Comment by danic85** [comment]
Further refinement.

Re-oriented the I2C connector based on advice from this article:
https://hackaday.com/2022/05/04/the-connector-zoo-i2c-ecosystems/

Also adjusted the other 3 pin JST connectors so that the power is in the center pin. This means that if the orientation is changed accidentally it has less chance of damaging the component as only the ground and GPIO will be swapped.  

Once the parts have arrived I will do some final tests then order the PCBs

![Screenshot from 2023-11-16 19-00-32](https://github.com/makerforgetech/modular-biped/assets/6583012/3d25c18d-73e9-430a-9b86-9f3d4dfd4eeb)
![Screenshot from 2023-11-16 18-59-53](https://github.com/makerforgetech/modular-biped/assets/6583012/81600237-1670-450f-bc59-3af3823b163d)

**Comment by danic85** [comment]
Checked all components against PCB and confirmed correct configuration. Swapped TX and RX into LLC to match text on module.

Dumping some datasheets and references here because I'm tired of googling them!

https://pinout.xyz/
https://cdn-shop.adafruit.com/product-files/2686/SK6812MINI_REV.01-1-2.pdf
https://learn.adafruit.com/adafruit-i2s-mems-microphone-breakout/raspberry-pi-wiring-test
https://learn.adafruit.com/adafruit-neodriver-i2c-to-neopixel-driver
https://learn.adafruit.com/adafruit-neodriver-i2c-to-neopixel-driver/downloads
https://cdn-learn.adafruit.com/downloads/pdf/adafruit-max98357-i2s-class-d-mono-amp.pdf
https://www.makerhero.com/img/files/download/RCWL0516-Datasheet.pdf

It seems like if I edit the PCB part reference, each time I import changes from the schematic it resets those values, so I'll change them in the schematic next time...

Latest screenshots:

![Screenshot from 2023-11-16 19-44-03](https://github.com/makerforgetech/modular-biped/assets/6583012/9cd59c15-49fd-42cc-b69f-13d72fa92c31)
![Screenshot from 2023-11-16 19-43-26](https://github.com/makerforgetech/modular-biped/assets/6583012/2033d9f2-bc90-4d94-b4f2-19c07529ad11)

**Comment by danic85** [comment]
Final update for today. Added a hole to feed the AMP output into an internal speaker, and some other small adjustments.
![Screenshot from 2023-11-16 21-22-09](https://github.com/makerforgetech/modular-biped/assets/6583012/6be840d6-8086-44a3-8129-85febc548e6b)
![Screenshot from 2023-11-16 21-21-51](https://github.com/makerforgetech/modular-biped/assets/6583012/1820e0f6-f280-47ac-912c-9a03611d378d)
![Screenshot from 2023-11-16 21-21-38](https://github.com/makerforgetech/modular-biped/assets/6583012/810808a4-9ae9-403d-899b-45c7678cdf3c)

**Comment by danic85** [comment]
Added 100nF 6030 decoupling capacitors to the SK6812 neopixels following recommendations on the [datasheet](https://cdn-shop.adafruit.com/product-files/2686/SK6812MINI_REV.01-1-2.pdf) and [this post](https://electronics.stackexchange.com/questions/548666/individual-ws2812b-circuit-do-i-need-a-capacitor-power-distribution).

Checked everything over and placed an order for 5 to test once they arrive. 

![Screenshot from 2023-11-17 20-50-43](https://github.com/makerforgetech/modular-biped/assets/6583012/f2cd802d-78b4-4da0-b722-29b35c9fd696)
![Screenshot from 2023-11-17 20-47-26](https://github.com/makerforgetech/modular-biped/assets/6583012/b373e570-7273-433d-a579-b497f3907beb)
![Screenshot from 2023-11-17 20-47-05](https://github.com/makerforgetech/modular-biped/assets/6583012/0c794189-0299-4263-9bb1-6c88f8aae8ee)
![Screenshot from 2023-11-17 20-46-31](https://github.com/makerforgetech/modular-biped/assets/6583012/81a93bf0-3b99-493e-bb34-bf3b2eefc7ba)
![Screenshot from 2023-11-17 20-46-07](https://github.com/makerforgetech/modular-biped/assets/6583012/6001ec49-7a45-404e-bf8b-1dd102c75c59)

**Comment by danic85** [comment]
New video on YouTube and article on the website. 

https://www.makerforge.tech/posts/kicad-basics/

https://youtu.be/0in5KH4QHYs

**Comment by danic85** [comment]
Just a quick update to say that this PCB has arrived and I'm beginning testing. The neopixels work with the i2c adapter, I want to test the amp and mic next as that was complex to get working in the previous version.

**Reply by bill1289** [reply]
Dan, you are very ambitious and driven. How did the testing go?

**Reply by danic85** [reply]
Mostly fine... I made a couple of videos on the build on YouTube. The audio is still a problem because of the Pi5 support issues, but the PCB works :laughing: 

I'm planning on doing another walkthrough to talk about the build and latest challenges, so if there's anything you're interested in hearing about let me know.

**Comment by RoboTech-1** [comment]
@danic85 Can you share the gerber file of your latest custom PCB

**Reply by danic85** [reply]
The files for the pi 5 pcb is available in the repo (check the wiki). The CM5 PCB is not released yet. Thanks!

**Comment by RoboTech-1** [comment]
@danic85 How i can connect with you ?
